src/main/java/com/conveyal/analysis/controllers/WorkerProxyController.java

```diff
@@ -41,59 +48,93 @@ public void registerEndpoints (Service sparkService) {
         // Ideally we'd have region, project, bundle, and worker version parameters.
         // Those could be path parameters, x-conveyal-headers, etc.
         // Path starts with api to ensure authentication
-        sparkService.get("/api/bundles/:bundle/workerProxy/:workerVersion", this::proxyGet);
+        sparkService.path("/api/workers/:workerVersion/bundles/:bundleId", () -> {
+            sparkService.get("", this::getWorkerStatus, JsonUtil.toJson);
+            sparkService.get("/*", this::proxyRequest);
+        });
     }
 
     /**
-     * Handler for GET requests to be proxied to a worker. This endpoint is contacted by the frontend.
-     * These requests typically come from an interactive session where the user is using the web UI.
-     * @return whatever the worker responds, usually an input stream. Spark serializer chain can properly handle streams.
+     * Get the address of an existing worker or create a new worker and return `null`.
      */
-    private Object proxyGet (Request request, Response response) {
-
-        final long startTimeMsec = System.currentTimeMillis();
-
-        final String bundleId = request.params("bundleId");
-        final String workerVersion = request.params("workerVersion");
-
-        WorkerCategory workerCategory = new WorkerCategory(bundleId, workerVersion);
+    private String getOrStartWorker (String bundleId, String workerVersion, UserPermissions userPermissions) {
```

Comment:
This code is mostly a duplicate of the code in the `BrokerController` (in fact it was originally copied from there). Should we abstract this out as a helper method in the broker itself?

```java
   /**
     * Get or start worker.
     */
    public synchronized String getOrStartWorker(WorkerCategory workerCategory, WorkerTags workerTags) {
        String address = getWorkerAddress(workerCategory);
        if (address == null) {
            // There are no workers that can handle this request. Request some.
            createOnDemandWorkerInCategory(workerCategory, workerTags);
        } else {
            // Workers exist in this category, clear out any record that we're waiting for one to start up.
            // FIXME the tracking of which workers are starting up should really be encapsulated using a "start up if needed" method.
            recentlyRequestedWorkers.remove(workerCategory);
        }
        return address;
    }
```

At the very least, we shouldn't leave the replicated `FIXME` in two different places.

Reply:
I definitely agree, this should be factored out into a shared method. In fact, as a best practice we should avoid factoring out this much code even when making a prototype because it then requires us to spot and re-merge the duplicates months down the line.